src/serac/physics/materials/tests/thermomechanical_material.cpp

```diff
@@ -0,0 +1,303 @@
+#include <iostream>
+
+#include <gtest/gtest.h>
+
+#include "serac/numerics/functional/tensor.hpp"
+#include "serac/numerics/functional/tuple.hpp"
+#include "serac/numerics/functional/tuple_arithmetic.hpp"
+
+
+namespace serac {
+
+static constexpr auto I = Identity<3>();
+
+// An arbitrary rotation tensor for frame indifference tests
+static const tensor<double, 3, 3> Q{{{ 0.852932108456964,  0.405577416697288, -0.328654495524269 },
+                                     {-0.51876824611259 ,  0.728725283562492, -0.44703351990878  },
+                                     { 0.058192140263313,  0.551784758906877,  0.831953877717782 }}};
+
+/**
+ * @brief Compute Green's strain from the displacement gradient
+ */
+template <typename T>
+auto green_strain(const tensor<T, 3, 3>& grad_u)
+{
+  return 0.5*(grad_u + transpose(grad_u) + transpose(grad_u)*grad_u);
+}
+
+
+/// @brief Green-Saint Venant isotropic thermoelastic model
+struct   ThermoelasticMaterial {
+  double E;          ///< Young's modulus
+  double nu;         ///< Poisson's ratio
+  double C;          ///< volumetric heat capacity
+  double alpha;      ///< thermal expansion coefficient
+  double theta_ref;  ///< datum temperature
+  double k;          ///< thermal conductivity
+
+  struct State {/* this material has no internal variables */};
+
+  /**
+   * @brief Evalute constitutive variables for thermomechanics
+   *
+   * @param[in] grad_u displacement gradient
+   * @param[in] theta temperature
+   * @param[in] grad_theta temperature gradient
+   * @param[in] grad_u_old displacement gradient at previous time step
+   * @param[in] theta_old temperature at previous time step
+   * @param[in] dt time increment
+   * @param[out] P Piola stress
+   * @param[out] cv0 volumetric heat capacity in ref config
+   * @param[out] s0 internal heat supply in ref config
+   * @param[out] q0 Piola heat flux
+   */
+  template <typename T1, typename T2, typename T3>
+  auto calculate_constitutive_outputs(const tensor<T1, 3, 3>& grad_u,
+                                      T2 theta,
+                                      const tensor<T3, 3>& grad_theta,
+                                      State& /*state*/,
+                                      const tensor<double, 3, 3>& grad_u_old,
+                                      double /*theta_old*/,
+                                      double dt)
```

Comment:
@jamiebramwell I was thinking it would be good to change the other materials to follow this pattern, where instead of having the entire argument type be an arbitrary type, the arbitrary-ness is only in what is stored in the tensor. e.g.
```cpp
const tensor<T1, 3, 3>& grad_u
```
instead of
```cpp
const T1 & grad_u
```
This way, the function signature reveals more about the shapes of each argument.

Reply:
I agree - this is nicer.